src/NHibernate.Test/Futures/QuerytBatchFixture.cs

```diff
@@ -0,0 +1,466 @@
+			using (var session = OpenSession())
+			{
+				var future1 = session.QueryOver<EntityComplex>()
+						.Where(x => x.Version >= 0)
+						.TransformUsing(new ListTransformerToInt()).Future<int>();
+
+				var future2 = session.Query<EntityComplex>().Where(ec => ec.Version > 2).ToFuture();
+				var future3 = session.Query<EntitySimpleChild>().Select(sc => sc.Name).ToFuture();
+
+				var future4 = session
+						.Query<EntitySimpleChild>()
+						.ToFutureValue(sc => sc.FirstOrDefault());
+
+				Assert.That(future1.GetEnumerable().Count(), Is.GreaterThan(0), "Empty results are not expected");
+				Assert.That(future2.GetEnumerable().Count(), Is.EqualTo(0), "This query should not return results");
+				Assert.That(future3.GetEnumerable().Count(), Is.GreaterThan(1), "Empty results are not expected");
+				Assert.That(future4.Value, Is.Not.Null, "Loaded entity should not be null");
+
+				if (SupportsMultipleQueries)
+					Assert.That(sqlLog.Appender.GetEvents().Length, Is.EqualTo(1));
+			}
+		}
+
+		[Test]
+		public void CanCombineCriteriaAndHqlInBatch()
+		{
+			using (var session = OpenSession())
+			{
+				var batch = NewBatch(session);
+
+				var futureBatch1 =
+					batch.AddAsList<int>(
+					session.QueryOver<EntityComplex>()
+							.Where(x => x.Version >= 0)
+							.TransformUsing(new ListTransformerToInt()));
+
+				var futureEntComplList = batch.AddAsList(session.QueryOver<EntityComplex>().Where(x => x.Version >= 1));
+
+				var futureList3 = batch.AddAsList(session.Query<EntityComplex>().Where(ec => ec.Version > 2));
+
+				using (var sqlLog = new SqlLogSpy())
+				{
+					IList<int> list1 = futureBatch1.Value;
+					IList<EntityComplex> list2 = futureEntComplList.Value;
+					IList<EntityComplex> list3 = futureList3.Value;
+					if (SupportsMultipleQueries)
+						Assert.That(sqlLog.Appender.GetEvents().Length, Is.EqualTo(1));
+				}
+			}
+		}
+
+		[Test]
+		public void CanFetchCollectionInBatch()
+		{
+			using (var sqlLog = new SqlLogSpy())
+			using (var session = OpenSession())
+			{
+				var batch = NewBatch(session);
+
+				var q1 = session.QueryOver<EntityComplex>()
+								.Where(x => x.Version >= 0);
+
+				batch.Add(new CriteriaBatchItem<object>(q1.RootCriteria));
+
+				batch.Add(new LinqBatchItem<EntityComplex>(session.Query<EntityComplex>().Fetch(c => c.ChildrenList)));
+				batch.Execute();
+				var parent = session.Load<EntityComplex>(_parentId);
+				Assert.That(NHibernateUtil.IsInitialized(parent), Is.True);
+				Assert.That(NHibernateUtil.IsInitialized(parent.ChildrenList), Is.True);
+				if (SupportsMultipleQueries)
+					Assert.That(sqlLog.Appender.GetEvents().Length, Is.EqualTo(1));
+			}
+		}
+
+		[Test]
+		public void OnAfterLoad()
+		{
+			using (var session = OpenSession())
+			{
+				var batch = NewBatch(session);
+				IList<EntityComplex> results = null;
+				batch.AddOnAfterLoad(session.Query<EntityComplex>().WithOptions(o => o.SetCacheable(true)), list => results = list);
```

Comment:
>Or do you consider it could already be directly used as a multi-criteria/multi-query alternative

Yes I think it should be available as separate batcher. I see it as `IFuture...` interfaces as a modern NHibernate suggested approach for operating with batch results. So those extensions are not hard-wired to `session.FutureBatch`. It's also a suggested methods to use with directly created batches -  It's better to have the same approach everywhere. But if user has some custom batch implementations and so he doesn't use IFuture interfaces - I've added OnAfterLoad event to simplify integration.

I'm not sure if we can simply obsolete old multi interfaces and implementations - it could require a considerable amount of work to rewrite existing code to use new batcher. As quite different approach is used for retrieving results. So it might be a showstopper to upgrade for many projects.

Reply:
Keeping two different implementations for achieving the same functionality is a burden, and it may also confuse users about what they should use.

From my point of view, futures are an additional abstraction layer on top of multi-queries. It makes sens to let users use directly multi-queries/query-batch if they wish too.

But having two implementations of multi-queries/query-batch is not a thing which should be kept in my opinion.

> As quite different approach is used for retrieving results.

That is a point to fix before obsoleting the previous implementation.

But since this PR is already quite broad, we would have better to handle obsoleting the previous implementation in a dedicated PR.